### PR TITLE
Migrate state mutations (block/uncle rewards, dao) to params/mutations

### PR DESCRIPTION
**Pull request by iquidus**
Introduces a new "mutations" module (found in params/mutations/*), intended for state mutations that take place outside of standard transactions, e.g block/uncle rewards, dao.

This logic has traditionally been located within the consensus/ethash and consensus/misc modules. Given these state mutations are likely to vary between chains (And are not technically part of ethash) it makes more sense to keep them in params.

No changes have been made to chain configs. 

reference #20

### CODE REVIEW

**Review by meowsbits — COMMENTED** [review]
I'm not sure the `--mainnet` flag commit belongs in this branch, or? 

https://github.com/etclabscore/core-geth/pull/335/commits/4ed1399fd4f577be676fb06d2ddffe2e0032cddf

**Comment by iquidus** [comment]
> I'm not sure the `--mainnet` flag commit belongs in this branch, or?
> 
> [4ed1399](https://github.com/etclabscore/core-geth/commit/4ed1399fd4f577be676fb06d2ddffe2e0032cddf)

I have no idea what you are talking about ;)

Was a mistake, didn't notice my master wasn't clean before branching. Removed.

**Review by meowsbits — APPROVED** [review]
Looks fine to me. Per a team chat yesterday, there may be some unit tests for rewards in the works. Up to you @iquidus if you want to include those in the PR or a later one.

**Comment by ziogaschr** [comment]
LGTM too! Nice decoupling

**Comment by meowsbits** [comment]
@iquidus Do you intend this to be ready for re-review/merge at this point?

**Review by meowsbits — CHANGES_REQUESTED** [review]
Eventually I'd suggest the commit message https://github.com/etclabscore/core-geth/pull/335/commits/41fa2e43a950a0df6b5a34ada81db8044d3e65cf include more information.

File: params/mutations/rewards_test.go
Comment (iquidus) [inline params/mutations/rewards_test.go]: This test (from the old code), still needs resolving. It runs (after refactor), but fails massively, further study required.

File: params/mutations/rewards_test.go
Comment (iquidus) [inline params/mutations/rewards_test.go]: duplicate tests to a degree, however I figure more never hurts (ported over from core-pool).

File: params/mutations/rewards_test.go
Comment (iquidus) [reply params/mutations/rewards_test.go]: resolved.

**Comment by iquidus** [comment]
typo in commit message c3b1080

**Comment by meowsbits** [comment]
This is looking good! 

What do you think about creating an additional [regression test](https://github.com/etclabscore/core-geth/tree/master/tests/regression/simulated) to have a full end-to-end test focusing on block rewards? 

Indeed, there are tests that touch and handle rewards (resultant state states), but I don't believe they cover some important configs (which is what I'm always worried about...:wink:) 

Like test a "condensed" ETC config with ECIP1017 Era Rounds of `100` or something with a chain of length 100, with some random uncles in there for good measure?

In your opinion, would this extra test warranted for inclusion at this PR, or would you rather have something like that as a follow up?

**Comment by meowsbits** [comment]
Ah, maybe I am on tracks I have already walked. https://github.com/etclabscore/core-geth/blob/master/tests/regression/simulated/classic-condense-state/classic.conf.json#L32

This one linked has ECIP1017 eras of `500`, and a total of 1467 blocks, which takes the test into Era 3. I don't think there's any uncles in there, though.

**Review by meowsbits — APPROVED** [review]

**Comment by meowsbits** [comment]
Just waiting for some moar tests to finish locally.

**Comment by meowsbits** [comment]
@iquidus Any idea why I'm getting this on this branch locally?

```
make all
make[1]: Entering directory '/home/ia/go/src/github.com/ethereum/go-ethereum'
env GO111MODULE=on go run build/ci.go install
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/abidump ./cmd/abidump
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/abigen ./cmd/abigen
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/ancient-store-mem ./cmd/ancient-store-mem
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/bootnode ./cmd/bootnode
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/checkpoint-admin ./cmd/checkpoint-admin
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/clef ./cmd/clef
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/devp2p ./cmd/devp2p
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/echaindb ./cmd/echaindb
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/echainspec ./cmd/echainspec
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/ethkey ./cmd/ethkey
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/evm ./cmd/evm
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/faucet ./cmd/faucet
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/geth ./cmd/geth
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/p2psim ./cmd/p2psim
>>> /home/ia/go1.16.2.linux-amd64/bin/go build -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -v -o /home/ia/go/src/github.com/ethereum/go-ethereum/build/bin/rlpdump ./cmd/rlpdump
make[1]: Leaving directory '/home/ia/go/src/github.com/ethereum/go-ethereum'
make test
make[1]: Entering directory '/home/ia/go/src/github.com/ethereum/go-ethereum'
env GO111MODULE=on go run build/ci.go test
>>> /home/ia/go1.16.2.linux-amd64/bin/go test -ldflags -X main.gitCommit=410cb0df3b3da9b1c7ae22a1100822bbe04e2e96 -X main.gitDate=20210410 -p 1 ./...
?       github.com/ethereum/go-ethereum [no test files]
ok      github.com/ethereum/go-ethereum/accounts        (cached)
ok      github.com/ethereum/go-ethereum/accounts/abi    (cached)
--- FAIL: TestGolangBindings (0.14s)
    bind_test.go:1758: failed to run binding test: exit status 1
        callbackparam.go:10:2: module github.com/ethereum/go-ethereum provides package github.com/ethereum/go-ethereum and is replaced but not required; to add it:
                go get github.com/ethereum/go-ethereum
        callbackparam.go:11:2: module github.com/ethereum/go-ethereum provides package github.com/ethereum/go-ethereum/accounts/abi and is replaced but not required; to add it:
                go get github.com/ethereum/go-ethereum
        callbackparam.go:12:2: module github.com/ethereum/go-ethereum provides package github.com/ethereum/go-ethereum/accounts/abi/bind and is replaced but not required; to add it:
                go get github.com/ethereum/go-ethereum
        callbackparam.go:13:2: module github.com/ethereum/go-ethereum provides package github.com/ethereum/go-ethereum/common and is replaced but not required; to add it:
                go get github.com/ethereum/go-ethereum
        callbackparam.go:14:2: module github.com/ethereum/go-ethereum provides package github.com/ethereum/go-ethereum/core/types and is replaced but not required; to add it:
                go get github.com/ethereum/go-ethereum
        callbackparam.go:15:2: module github.com/ethereum/go-ethereum provides package github.com/ethereum/go-ethereum/event and is replaced but not required; to add it:
                go get github.com/ethereum/go-ethereum
FAIL
FAIL    github.com/ethereum/go-ethereum/accounts/abi/bind       0.170s
ok      github.com/ethereum/go-ethereum/accounts/abi/bind/backends      (cached)
?       github.com/ethereum/go-ethereum/accounts/external       [no test files]
^Csignal: interrupt
Makefile:40: recipe for target 'test' failed
make[1]: *** [test] Error 1
.ia/Makefile:8: recipe for target 'check_simple' failed
make: *** [check_simple] Interrupt
```

**Comment by meowsbits** [comment]
This problem goes away once merged with `master`.

**Review by meowsbits — APPROVED** [review]

**Comment by iquidus** [comment]
> --- FAIL: TestGolangBindings (0.14s)

iirc I encountered this with go 1.16, i reverted back to 1.15 (with intentions to look at it later), however i believe it was fixed in master.

**Comment by meowsbits** [comment]
@ziogaschr PTAL and review just for a second opinion, since this is consensus-sensitive code.

**Comment by iquidus** [comment]
classic mainnet sync: complete.
ethereum mainnet sync: in progress.

**Review by ziogaschr — APPROVED** [review]
LGTM. I am waiting for sync to finish on classic on my side too.

**Comment by iquidus** [comment]
ethereum mainnet sync: sync'd up to block 12 244 000 (berlin activation)

**Review by meowsbits — CHANGES_REQUESTED** [review]
Just giving a red review to remind that this PR has `master` as a base branch; I think it should want `rc1` as the base now.

**Comment by iquidus** [comment]
note: there's a failing openrpc test, inherited from the rc-1 branch.

**Comment by meowsbits** [comment]
> note: there's a failing openrpc test, inherited from the rc-1 branch.

@iquidus I'm not sure I see the failing test you're referencing? 
The `push` build at the `rc1` branch is passing https://travis-ci.org/github/etclabscore/core-geth/builds/770798087, and the failing `pr` build here looks to be the fault of timing out on the H-Z step https://travis-ci.org/github/etclabscore/core-geth/jobs/771926408

Can you give instructions for reproduction, or source where you're seeing the test failure?

**Comment by iquidus** [comment]
the openrpc failure can be ignored. This was purely a local issue. 

Docker (potentially hung from earlier tests), still had port 8545 occupied (tcp6 - LISTENING). Killing docker, freeing the port, has resolved the failure.

**Comment by iquidus** [comment]
merged into #376

**Pull request closed without merging**
